pkg/storegateway/bucket_index_reader.go

```diff
@@ -32,18 +32,35 @@ import (
 
 // expandedPostingsPromise is the promise returned by bucketIndexReader.expandedPostingsPromise.
 // The second return value indicates whether the returned data comes from the cache.
-type expandedPostingsPromise func(ctx context.Context) ([]storage.SeriesRef, bool, error)
+type expandedPostingsPromise func(ctx context.Context) ([]storage.SeriesRef, []*labels.Matcher, bool, error)
+
+type postingsSelectionStrategy interface {
+	name() string
+	selectPostings([]postingGroup) (selected, omitted []postingGroup)
```

Comment:
(Opinionated): I'd `s/selectPostings/select/`: we're not really selecting postings, but groups, and anyway the function signature already tells us the type, and it's cleaner.

Reply:
I see that we can't have `select` as a keyword, so maybe `choose`? Nevermind, ignore this.